src/main/java/educonnect/model/student/timetable/AvailableSlots.java

```diff
@@ -0,0 +1,199 @@
+        if (days.containsKey(dayOfWeek)) {
+            days.get(dayOfWeek).addPeriod(period);
+        } else {
+            days.put(dayOfWeek, new Day(dayOfWeek, new ArrayList<>(List.of(period)), false));
+        }
+    }
+
+    /**
+     * Checks if the {@code AvailableSlots} object has any common slots found.
+     *
+     * @return {@code true} if there is any {@code Period} found.
+     */
+    public boolean hasCommonSlots() {
+        if (this.days.isEmpty()) {
+            return false;
+        }
+
+        for (Day day: this.days.values()) {
+            if (day.hasPeriods()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Finds all common available slots, given each {@code AvailableSlots} from each {@code Student}.
+     *
+     * @param allAvailableSlots {@code ArrayList} of {@code AvailableSlots} objects
+     *     generated from {@code Student} objects.
+     * @return an {@code AvailableSlot} object containing all common slots from multiple students.
+     */
+    public static AvailableSlots findAllCommonSlots(ArrayList<AvailableSlots> allAvailableSlots) {
+        AvailableSlots result = new AvailableSlots(new HashSet<>());
+        ArrayList<DayOfWeek> allDaysPossible = findAllDays(allAvailableSlots);
+
+        for (DayOfWeek dayOfWeek : allDaysPossible) {
+            ArrayList<Day> allDays = new ArrayList<>();
+
+            for (AvailableSlots eachSlot : allAvailableSlots) {
+                allDays.add(eachSlot.getDay(dayOfWeek));
+            }
+            Day day = Day.findAllCommonSlots(allDays);
+            result.days.put(day.getDayOfWeek(), day);
+        }
+        result.setCommonSlots();
+        return result;
+    }
+
+    /**
+     * Finds all available {@code DayOfWeek} across all {@code AvailableSlots} objects.
+     *
+     * @param allAvailableSlots {@code ArrayList} of {@code AvailableSlots} objects
+     *     generated from {@code Student} objects.
+     * @return a list of DayOfWeek objects that appears in any of the {@code AvailableSlots}.
+     */
+    static ArrayList<DayOfWeek> findAllDays(ArrayList<AvailableSlots> allAvailableSlots) {
+        if (allAvailableSlots.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        Set<DayOfWeek> allDays = new HashSet<>(allAvailableSlots.get(0).days.keySet());
+
+        for (AvailableSlots eachSlot : allAvailableSlots) {
+            allDays.retainAll(eachSlot.days.keySet());
+        }
+
+        ArrayList<DayOfWeek> allDayArrayList = new ArrayList<>(allDays);
+        Collections.sort(allDayArrayList);
+        return allDayArrayList;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+
+        result = prime * result + this.days.hashCode();
+        result = prime * result + (this.isCommonSlots ? prime : 0);
+        return result;
+    }
```

Comment:
This is important as it is used in HashMap. Good implementation.
One thing I would change is `(this.isCommonSlots ? anotherPrime : 0)`,
as adding `prime` is basically `prime * (result + 1)`. I'm not sure if it would lead to less collisions though.